content/blog/2019-10-28-self/index.md

```diff
@@ -0,0 +1,176 @@
+at the famed Xerox PARC labs takes the philosophy of dynamism to a logical
+extreme -- *everything* in self is a message to an object. This includes
+Java-like method calls on object *as well as* control structures like loops
+and conditionals.
+
+For example, the Self condition `IfTrue:IfFalse:` is a method invocation on
+the boolean object `true` and `false`. This means that a simple conditional:
+
+```
+if (x) 1 else 2
+```
+
+in Self parlance (written with JavaScript-esque syntax) looks like this:
+
+
+```
+x.If({ true: () => 1, false: () => 2 })
+```
+
+The runtime behavior of this program is invoking the method `IfTrue` on the
+object `x`, which can be a `true` or `false` but is not required to be those,
+and execute the "thunk" (a function with no argument) corresponding to the true
+or the false branch.  Note that `x` is not required to be `true` or `false`.
+Any Self object can define the `If` method and specify its conditional
+behavior.
+
+Considering this dynamism, a Self compiler must:
+
+1. **Respect the semantics of method calls**. Restricting or specializing conditionals and primitives don't follow the spirit of the language.
+2. **Provide interactive speed**. Recompiling after a small change is not an option because Self is meant for rapid exploration in a programming environment.
+3. **Preserve stack traces**. Self supports extreme reflection and introspection. A programming environment must be debuggable.
+4. **Generate fast code**. Interpreting the whole language is probably too slow.
+
+## The Solution
+
+While the paper goes into the nitty-gritty of object layouts and method
+invocations, the essence of the paper can be summarized as:
+
+>  When type information for an object is available, generate specialized code
+>  and let inlining and compiler optimizations work their magic.
+
+The ideas described in the paper have influenced the design of modern JITs for
+JavaScript which use a lot of [similar features][jit] to improve JavaScript execution
+times.
+
+### Object Layout
+
+Self programs use object prototypes to describe inheritance relationships.
+Unlike classes, which have _constructors_ used to build _instances_, Self uses
+prototypes, which act as _exemplars_ for other objects. Creating a new object
+from a prototype is as simple as _cloning_ it and setting its parent pointer to
+the prototype. Changing the common behavior of all clones is as simple as
+changing the behavior of a prototype. Since method and field lookups traverse
+the parent hierarchy, clones can also override methods and fields of their
+prototypes.
+
+A naive layout scheme for objects would copy all fields from a prototype and
+end up wasting a lot of space describing potentially shared behaviors. The Self
+compiler minimizes space usage of _clones_ derived from the same prototype by
+using _clone families_. A cloned object only stores its modifiable in its
+object and points to the clone family for its prototype. If the instance ever
```

Comment:
I'm not 100% sure what "in its object" means—maybe introduce something like an "allocation" to mean the actual memory for laying out fields, as opposed to "object" which is more abstract?